### DISCUSSION TITLE
Use WinAppDriver to automate application testing

### DISCUSSION THREAD

**Discussion by LucasYao93** [AvaloniaUI/Avalonia#16859] [Q&A]
Can WinAppDriver be used to automate application testing?

WinAppDriver: https://github.com/microsoft/WinAppDriver?tab=readme-ov-file

**Comment by stevemonaco** [comment]
Avalonia does integration testing using Appium, particularly to test native integrations with OS concerns such as windowing, popups, tray, etc.

If your app doesn't require that, then it's strongly suggested to use the Headless testing support Avalonia has. The reason being that crossplatform integration testing is a mess: WinAppDriver was abandoned by Microsoft years ago. There's no good solution for Linux or mobile that Appium supports. FlaUI works, but only for Windows. FlaUI.WebDriver isn't quite ready to replace WinAppDriver for integration testing via Appium on Windows.
